--- portaudio/src/hostapi/alsa/pa_linux_alsa.c.orig	2011-03-05 14:01:39.170254000 -0500
+++ portaudio/src/hostapi/alsa/pa_linux_alsa.c	2011-03-05 14:32:01.121718000 -0500
@@ -716,12 +716,9 @@
     return (const PaAlsaDeviceInfo *)hostApi->deviceInfos[device];
 }
 
-/** Uncommented because AlsaErrorHandler is unused for anything good yet. If AlsaErrorHandler is
-    to be used, do not forget to register this callback in PaAlsa_Initialize, and unregister in Terminate.
-*/
-/*static void AlsaErrorHandler(const char *file, int line, const char *function, int err, const char *fmt, ...)
+static void AlsaErrorHandler(const char *file, int line, const char *function, int err, const char *fmt, ...)
 {
-}*/
+}
 
 PaError PaAlsa_Initialize( PaUtilHostApiRepresentation **hostApi, PaHostApiIndex hostApiIndex )
 {
@@ -746,10 +743,7 @@
     (*hostApi)->OpenStream = OpenStream;
     (*hostApi)->IsFormatSupported = IsFormatSupported;
 
-    /** If AlsaErrorHandler is to be used, do not forget to unregister callback pointer in
-        Terminate function.
-    */
-    /*ENSURE_( snd_lib_error_set_handler(AlsaErrorHandler), paUnanticipatedHostError );*/
+    ENSURE_( snd_lib_error_set_handler(AlsaErrorHandler), paUnanticipatedHostError );
 
     PA_ENSURE( BuildDeviceList( alsaHostApi ) );
 
@@ -796,9 +790,7 @@
 
     assert( hostApi );
 
-    /** See AlsaErrorHandler and PaAlsa_Initialize for details.
-    */
-    /*snd_lib_error_set_handler(NULL);*/
+    snd_lib_error_set_handler(NULL);
 
     if( alsaHostApi->allocations )
     {
