--- portaudio/configure.orig	2010-03-16 08:23:13.000000000 -0400
+++ portaudio/configure	2010-03-31 08:37:27.000000000 -0400
@@ -19448,10 +19448,10 @@
         if test "x$enable_mac_universal" = "xyes" ; then
            mac_version_min="-mmacosx-version-min=10.3"
            if [ -d /Developer/SDKs/MacOSX10.5.sdk ] ; then
-              mac_arches="-arch i386 -arch ppc -arch x86_64 -arch ppc64"
+              mac_arches="-arch ppc"
               mac_sysroot="-isysroot /Developer/SDKs/MacOSX10.5.sdk"
            else
-              mac_arches="-arch i386 -arch ppc"
+              mac_arches="-arch ppc"
               mac_sysroot="-isysroot /Developer/SDKs/MacOSX10.4u.sdk"
            fi
         else
