Index: Slim/Web/HTTP.pm
===================================================================
--- Slim/Web/HTTP.pm	(revision 28009)
+++ Slim/Web/HTTP.pm	(revision 28008)
@@ -2176,11 +2176,11 @@
 	my $client     = shift;
 	my $httpClient = shift;
 
-	if ( $httpClient == $client->streamingsocket() ) {
-
+	if ( $httpClient->connected() ) {
+		
 		# Bug 10085 - This might be a callback for an old connection  
 		# which we decided to close after establishing the timer, so
-		# only kill the timer if we were called for the active streaming connection;
+		# only kill the timer if we were called for an active connection;
 		# otherwise we might kill the timer related to the next connection too.
 		Slim::Utils::Timers::killTimers($client, \&tryStreamingLater);
 		
