--- Slim/Web/HTTP.pm    (revision 30371)
+++ Slim/Web/HTTP.pm    (working copy)
@@ -2214,9 +2214,10 @@
                # only kill the timer if we were called for the active streaming connection;
                # otherwise we might kill the timer related to the next connection too.
                Slim::Utils::Timers::killTimers($client, \&tryStreamingLater);
-
-               Slim::Networking::Select::addWrite($httpClient, \&sendStreamingResponse, 1);
        }
+
+       # Always call sendStreamingResponse so we ensure the socket gets closed
+       Slim::Networking::Select::addWrite($httpClient, \&sendStreamingResponse, 1);
 }

